docs/design/2022-03-24-get-lock.md

```diff
@@ -0,0 +1,44 @@
+# Proposal: Support `GET LOCK`
+
+- Author(s): [morgo](https://github.com/morgo)
+- Tracking Issue: https://github.com/pingcap/tidb/issues/14994
+
+## Abstract
+
+This proposes an implementation of `GET_LOCK()` with some limitations.
+
+## Background
+
+MySQL supports advisory locks (aka `GET_LOCK()`) as a way of serializing access to a resource. This is commonly used by applications such as ORMs to ensure that only one process can perform DDL at a time.
+
+Advisory locks are session scoped (not transaction) locks. Up until MySQL 5.7 acquiring any new lock would free all previously acquired locks. This was because the locking system did not support deadlock detection internally.
+
+## Proposal
+
+This proposes that we use an internal table (`mysql.advisory_locks`) with the `lockName` as the `PRIMARY KEY`. When a user acquires a lock with `GET_LOCK()`, we insert a row into the table as part of an internal pessimistic transaction. The internal transaction is never committed, and stays open until `RELEASE_LOCK()` is called, or the session is closed.
```

Comment:
This means that the lock gets removed if the network connection is lost or if the `tidb-server` process restarts.

Reply:
Yes. But I believe TiDB has previously had some issues here.

Reply:
> But I believe TiDB has previously had some issues here.

I believe everything is fine now. All remaining locks will be cleaned automatically.

Reply:
It turns out there is/was a bug here. Discovered by QA today: https://github.com/pingcap/tidb/pull/34757